components/script/window_named_properties.rs

```diff
@@ -0,0 +1,263 @@
+    jsid, GetPropertyKeys, GetWellKnownSymbol, JS_SetImmutablePrototype, SymbolCode, JSITER_HIDDEN,
+    JSITER_OWNONLY, JSITER_SYMBOLS,
+};
+use js::jsapi::{
+    Handle, HandleObject, JSClass, JSContext, JSErrNum, MutableHandleObject, UndefinedHandleValue,
+};
+use js::jsapi::{
+    HandleId, JSClass_NON_NATIVE, MutableHandle, MutableHandleIdVector, ObjectOpResult,
+    PropertyDescriptor, ProxyClassExtension, ProxyClassOps, ProxyObjectOps,
+    JSCLASS_DELAY_METADATA_BUILDER, JSCLASS_IS_PROXY, JSCLASS_RESERVED_SLOTS_MASK,
+    JSCLASS_RESERVED_SLOTS_SHIFT,
+};
+use js::jsid::SymbolId;
+use js::jsval::UndefinedValue;
+use js::rust::IntoHandle;
+use js::rust::{Handle as RustHandle, MutableHandle as RustMutableHandle};
+use js::rust::{HandleObject as RustHandleObject, MutableHandleObject as RustMutableHandleObject};
+use libc;
+use std::ptr;
+
+struct SyncWrapper(*const libc::c_void);
+#[allow(unsafe_code)]
+unsafe impl Sync for SyncWrapper {}
+
+lazy_static! {
+    static ref HANDLER: SyncWrapper = {
+        let traps = ProxyTraps {
+            enter: None,
+            getOwnPropertyDescriptor: Some(get_own_property_descriptor),
+            defineProperty: Some(define_property),
+            ownPropertyKeys: Some(own_property_keys),
+            delete_: Some(delete),
+            enumerate: None,
+            getPrototypeIfOrdinary: Some(get_prototype_if_ordinary),
+            getPrototype: None,
+            setPrototype: None,
+            setImmutablePrototype: None,
+            preventExtensions: Some(prevent_extensions),
+            isExtensible: Some(is_extensible),
+            has: None,
+            get: None,
+            set: None,
+            call: None,
+            construct: None,
+            hasOwn: None,
+            getOwnEnumerablePropertyKeys: None,
+            nativeCall: None,
+            objectClassIs: None,
+            className: Some(class_name),
+            fun_toString: None,
+            boxedValue_unbox: None,
+            defaultValue: None,
+            trace: None,
+            finalize: None,
+            objectMoved: None,
+            isCallable: None,
+            isConstructor: None,
+        };
+
+        #[allow(unsafe_code)]
+        unsafe {
+            SyncWrapper(CreateProxyHandler(&traps, ptr::null()))
+        }
+    };
+}
+
+#[allow(unsafe_code)]
+unsafe extern "C" fn get_own_property_descriptor(
+    cx: *mut JSContext,
+    proxy: HandleObject,
+    id: HandleId,
+    desc: MutableHandle<PropertyDescriptor>,
+    is_none: *mut bool,
+) -> bool {
+    let cx = SafeJSContext::from_ptr(cx);
+
+    if id.is_symbol() {
+        if id.get().asBits_ == SymbolId(GetWellKnownSymbol(*cx, SymbolCode::toStringTag)).asBits_ {
+            rooted!(in(*cx) let mut rval = UndefinedValue());
+            "WindowProperties".to_jsval(*cx, rval.handle_mut());
+            set_property_descriptor(
+                RustMutableHandle::from_raw(desc),
+                rval.handle(),
+                0,
```

Comment:
This should be `JSPROP_READONLY`.

Reply:
done in 8fe6f28dca9cc2776768179b5abde6d4edf16184